Printrboard G2 support

Bob-the-Kuhn opened the pull request:
[Concise Diff](13116/files?w=1)

**This PR is being opened so that I can get feedback on the methods used.**  

Feedback on the code is also welcomed.

TO DO:
- Add support for the LCD module used by this board.
- Find a better way to implement the new port/pin pairs.

---

This PR add support for the G2 PRINTRBOARD. See Issue #13020.

This board is based on a SAM3X8**C** which is a subset of the SAM3X8**E** used by DUE.  The main difference is a smaller package which results in no useable C, D or E ports.

TOP LEVEL CHANGES:
- The G2 uses 8 port/pin pairs that are not available on the DUE.  I have not found a nice way of implementing them.  This PR has the easiest to understand method I've found.  It replaces some of the port/pin definitions in FASTIO.  Unfortunately only the WRITE and READ functions work.  Since the FASTIO functions that set pins as inputs/outputs don't work with this approach, there is an ugly pin init section added (in  an unusual but convenient location).
- The G2 has a safety feature that kills the FETs if a certain pin isn't pulsed regularly.  Code to do this was added to the inactivity routine.
- One of the PWM motor current pins can not be controlled by the PWM module.  Because of this I've adapted an early version of the LPC1768 PWM code to use the PWM controller to generate interrupts which then play with the pins.
- The G2 PWM code is written as a conditional replacement for digipot_init() and :digipot_current().
- The current support for separate X & Y pins for the PWM motor current is spotty.  I've corrected/added as needed.

The G2 board also uses a UART to communicate with a smart LCD module.

thinkyhead committed:
Printrboard G2 support

thinkyhead commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2143,20 +2143,14 @@ void Stepper::init() {
-  last_direction_bits = 0
-    | (INVERT_X_DIR ? _BV(X_AXIS) : 0)
-    | (INVERT_Y_DIR ? _BV(Y_AXIS) : 0)
-    | (INVERT_Z_DIR ? _BV(Z_AXIS) : 0);
```
This change was made with a lot of fanfare. Apparently the first direction change can fail if these aren't set according to their inverting.

Bob-the-Kuhn replied:
Definitely not an intended change.  I'll check for other unintended changes and back them all out later today.

thinkyhead commented:
Pretty clean looking overall. I'll review it more over the next day or two.

Bob-the-Kuhn commented:
No hurry.  

This thing uses a smart LCD/SD/WIFI module.  I'm hoping there is something already in place within Marlin or that the Printrboard software can be adapted.

Bob-the-Kuhn commented:
Changed the **_SET_OUTPUT** to use a modified PinDescription array.  This eliminated the custom handling of setting the new port/pin pairs as outputs.

Don't understand why but … just changing the name of the array in the standard macro didn't work for the new port/pin pairs.  I had to remove all the calls and explicitly do the register write.  The macro now is more verbose but it executes faster because the subroutine call overhead is gone.

Also changed all the "#if MOTHERBOARD = BOARD_PRINTRBOARD_G2" to #if MB(PRINTRBOARD_G2"

thinkyhead commented on Marlin/src/HAL/HAL_DUE/G2_PWM.h:
```diff
@@ -0,0 +1,193 @@
+
+  NVIC_EnableIRQ(PWM_IRQn);     // Enable interrupt handler
+  NVIC_SetPriority(PWM_IRQn, NVIC_EncodePriority(0, 10, 0));  // normal priority for PWM module (can stand some jitter on the Vref signals)
+}
```
These method definitions must be moved to a `.cpp` file. Check out the rework of the LCD code to see how we eliminated this type of header mis-use.

Bob-the-Kuhn replied:
Will do

thinkyhead replied:
Should be all set now!

thinkyhead commented:
Squashed, rebased, and moved C++ code out of header…
```
git fetch origin
git checkout G2-PR
git reset --hard origin/G2-PR
```

thinkyhead commented:
I'm guessing this still needs work, since the `work_table` and `active_table` described in the comments don't exist.

thinkyhead committed:
Printrboard G2 changes
